src/XRP/network-clients/web-socket-network-client.ts

```diff
@@ -0,0 +1,155 @@
+import WebSocket = require('ws')
+import { XrpError, XrpErrorType } from '../shared'
+import {
+  WebSocketRequestOptions,
+  WebSocketResponse,
+  WebSocketStatusResponse,
+  WebSocketTransactionResponse,
+} from '../shared/rippled-web-socket-schema'
+
+const sleep = (ms: number): Promise<void> => {
+  return new Promise((resolve) => setTimeout(resolve, ms))
+}
+
+/**
+ * A network client for interacting with the rippled WebSocket API.
+ * @see https://xrpl.org/rippled-api.html
+ */
+export default class WebSocketNetworkClient {
+  private readonly socket: WebSocket
+  private accountCallbacks: Map<
+    string,
+    (data: WebSocketTransactionResponse) => void
+  >
+  private messageCallbacks: Map<string, (data: WebSocketResponse) => void>
+  private waiting: Map<string, WebSocketResponse | undefined>
+
+  /**
+   * Create a new WebSocketNetworkClient.
+   *
+   * @param webSocketUrl The URL of the rippled node to query.
+   * @see https://xrpl.org/monitor-incoming-payments-with-websocket.html
+   */
+  public constructor(
+    webSocketUrl: string,
+    private handleErrorMessage: (message: string) => void,
+  ) {
+    this.socket = new WebSocket(webSocketUrl)
+    this.accountCallbacks = new Map()
+    this.waiting = new Map()
+
+    this.messageCallbacks = new Map()
+    this.messageCallbacks.set('response', (data: WebSocketResponse) => {
+      const dataStatusResponse = data as WebSocketStatusResponse
+      this.waiting.set(dataStatusResponse.id, data)
+    })
+    this.messageCallbacks.set('transaction', this.handleTransaction)
+
+    this.socket.addEventListener('message', (event) => {
+      const parsedData = JSON.parse(event.data) as WebSocketResponse
+      const dataType = parsedData.type
+      const callback = this.messageCallbacks.get(dataType)
+      if (callback) {
+        callback(parsedData)
+      } else {
+        this.handleErrorMessage(
+          'Received unhandlable message: ' + (event.data as string),
+        )
+      }
+    })
+
+    this.socket.addEventListener('close', (event) => {
+      this.handleErrorMessage(
+        'Web socket disconnected, ' + (event.reason as string),
+      )
+    })
+
+    this.socket.addEventListener('error', (event) => {
+      this.handleErrorMessage('Error: ' + (event as string))
+    })
+  }
+
+  /**
+   * Properly handles incoming transactions from the web socket.
+   *
+   * @param data The web socket response received from the web socket.
+   */
+  private handleTransaction = (data: WebSocketResponse) => {
```

Comment:
can we use a function declaration here instead? the sdk seems to prefer that for class functions

Reply:
I get an error if I do that, where I set it in `messageCallbacks`

Reply:
yeah, that sounds right. `this` behaves in kind of an interesting way in JS so you'll have to do

```
(data: WebSocketResponse) => this.handleTransaction(data)
```

since we seem to have chosen to go with function declarations, seems better to be standardized here